src/Tvheadend.cpp

```diff
@@ -367,6 +367,39 @@ PVR_ERROR CTvheadend::GetChannels(bool radio, kodi::addon::PVRChannelsResultSet&
   return PVR_ERROR_NO_ERROR;
 }
 
+PVR_ERROR CTvheadend::GetChannelStreamProperties(const kodi::addon::PVRChannel& channel, std::vector<kodi::addon::PVRStreamProperty>& properties)
+{
+  if (!Settings::GetInstance().GetStreamingHTTP())
+    return PVR_ERROR_NO_ERROR;
+
+  CLockObject lock(m_mutex);
+
+  auto it = m_channels.find(channel.GetUniqueId());
+  if (it == m_channels.end())
+    return PVR_ERROR_FAILED;
+
+  std::string url = Settings::GetInstance().GetUseHTTPS() ? "https://" : "http://";
+  const std::string usn = Settings::GetInstance().GetUsername();
+  const std::string pas = Settings::GetInstance().GetPassword();
+
+  if (!usn.empty() && !pas.empty())
+    url += usn + ":" + pas + "@";
+
+  url += Settings::GetInstance().GetHostname() + ":";
+  url += std::to_string(Settings::GetInstance().GetPortHTTP());
+  url += "/stream/channelid/";
```

Comment:
No need to calculate the basic parts of the URL yourself. We have already have a helper method for that.

```c++
std::string path = "/stream/channelid/" + std::to_string(it->first);
const std::string streamingProfile = Settings::GetInstance().GetStreamingProfile();
if (!streamingProfile.empty())
  path += "?profile=" + streamingProfile;

const std::string url = m_conn.GetWebURL("%s", path);
```

Reply:
Thanks for your patience! I understand now. The build failed after I applied the change suggesting to change the last line to
```cpp
const std::string url = m_conn->GetWebURL("%s", path);
``` 
Even with that amendment some (but not all) of the builds still fail though. I'm getting:
```cpp
/home/travis/build/kodi-pvr/pvr.hts/src/Tvheadend.cpp:386:51: error: cannot pass
      object of non-trivial type 'std::string' (aka 'basic_string<char>')
      through variadic method; call will abort at runtime [-Wnon-pod-varargs]
  const std::string url = m_conn->GetWebURL("%s", path);
```
I assume I need to `static_cast` something, but again, out of my depths, here. Sorry for the hassle.

Reply:
path.c_str()

Reply:
Thanks for the c++ crash course, @ksooo. Working again...  :-)